google_guest_agent/network/manager/systemd_networkd_linux.go

```diff
@@ -0,0 +1,156 @@
+
+	if err := json.Unmarshal([]byte(res.StdOut), &interfaceStatus); err != nil {
+		return false, fmt.Errorf("failed to unmarshal interface status: %v", err)
+	}
+
+	administrativeState, found := interfaceStatus["AdministrativeState"]
+	if !found {
+		return false, fmt.Errorf("could not determine interface state, AdministrativeState was not present")
+	}
+
+	return administrativeState == "configured", nil
+}
+
+// Setup sets up the non-primary network interfaces for systemd-networkd by writing
+// configuration files to /usr/lib/systemd/network.
+func (n systemdNetworkd) Setup(ctx context.Context, config *cfg.Sections, payload []metadata.NetworkInterfaces) error {
+	// Create a network configuration file with default configurations for each network interface.
+	googleInterfaces, googleIpv6Interfaces := InterfaceListsIpv4Ipv6(payload)
+	if err := configureIPv6(ctx, googleIpv6Interfaces, payload[0]); err != nil {
+		return fmt.Errorf("error configuring IPv6: %v", err)
+	}
+	for _, iface := range googleInterfaces {
+		logger.Debugf("write %s network config for %s", n.Name(), iface)
+
+		var config *os.File
+
+		var dhcp = "ipv4"
+		if slices.Contains(googleIpv6Interfaces, iface) {
+			dhcp = "yes"
+		}
+
+		// The number in beginning of the file name signifies its priority. We want this
+		// to take priority before any of the default configurations that systemd-networkd
+		// has set.
+		config, err := os.Create(fmt.Sprintf("%s/50-%s-google-guest-agent.network", n.ConfigDir, iface))
+		if err != nil {
+			return err
+		}
+		defer config.Close()
+
+		// Default, simple configuration for any network interface.
+		contents := []string{
+			googleComment,
+			guestAgentConfigIDKey,
+			"[Match]",
+			fmt.Sprintf("Name=%s", iface),
+			"",
+			"[Network]",
+			fmt.Sprintf("DHCP=%s", dhcp),
+		}
+		_, err = config.WriteString(strings.Join(contents, "\n"))
+		if err != nil {
+			return err
+		}
+	}
+	// Restart the systemd-networkd service.
+	logger.Debugf("restarting systemd-networkd")
+	return run.Quiet(ctx, "systemctl", "restart", "systemd-networkd.service")
+}
+
+// Rollback deletes the configuration files created by the agent for systemd-networkd.
+func (n systemdNetworkd) Rollback(ctx context.Context, payload []metadata.NetworkInterfaces) error {
+	logger.Infof("rolling back changes for %s", n.Name())
+	interfaces, err := interfaceNames(payload)
+	if err != nil {
+		return fmt.Errorf("failed to get list of interface names: %v", err)
+	}
+
+	for _, iface := range interfaces[1:] {
+		// Find expected files.
+		configFile := fmt.Sprintf("%s/50-%s-google-guest-agent.network", n.ConfigDir, iface)
+		logger.Debugf("checking for %s", configFile)
+		contents, err := os.ReadFile(configFile)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			} else {
+				return err
+			}
+		}
+
+		// Double check that the file has the guest-managed key before deleting it.
+		if strings.Contains(string(contents), guestAgentConfigIDKey) {
```

Comment:
this is a ini file, you can use `github.com/go-ini/ini` just like we are doing in `google_guest_agent/cfg/cfg.go`.